testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java

```diff
@@ -4305,6 +4305,34 @@ public void testMergeSubqueries()
         assertUpdate("DROP TABLE " + targetTable);
     }
 
+    @Test
+    public void testMergeNonNullableColumns()
+    {
+        skipTestUnless(hasBehavior(SUPPORTS_MERGE) && hasBehavior(SUPPORTS_NOT_NULL_CONSTRAINT));
+
+        String targetTable = "merge_non_nullable_target_" + randomTableSuffix();
+
+        assertUpdate(createTableForWrites(format("CREATE TABLE %s (nation_name VARCHAR, region_name VARCHAR NOT NULL)", targetTable)));
+
+        assertUpdate(format("INSERT INTO %s (nation_name, region_name) VALUES ('FRANCE', 'EUROPE'), ('ALGERIA', 'AFRICA'), ('GERMANY', 'EUROPE')", targetTable), 3);
+
+        // Show that updating using a null value fails
+        assertThatThrownBy(() -> computeActual(format("MERGE INTO %s t\n", targetTable) +
+                " USING (VALUES ('ALGERIA', 'AFRICA')) s(nation_name, region_name)\n" +
+                " ON (t.nation_name = s.nation_name)\n" +
+                " WHEN MATCHED THEN UPDATE SET region_name = NULL"))
+                .hasMessage("Assigning NULL to non-null MERGE target table column");
+
+        // Show that inserting using a null value fails
+        assertThatThrownBy(() -> computeActual(format("MERGE INTO %s t\n", targetTable) +
+                " USING (VALUES ('IMAGINARIA', 'AFRICA')) s(nation_name, region_name)\n" +
+                " ON (t.nation_name = s.nation_name)\n" +
+                " WHEN NOT MATCHED THEN INSERT (nation_name, region_name) VALUES ('IMAGINARIA', NULL)"))
+                .hasMessage("Assigning NULL to non-null MERGE target table column");
+
+        assertUpdate("DROP TABLE " + targetTable);
```

Comment:
Could you please add one more test when you insert null that is a result of an expression ? For example:
```
assertThatThrownBy(() -> computeActual(format("MERGE INTO %s t\n", targetTable) +
                " USING (VALUES ('ALGERIA', 'AFRICA')) s(nation_name, region_name)\n" +
                " ON (t.nation_name = s.nation_name)\n" +
                " WHEN MATCHED THEN UPDATE SET region_name = TRY(5/0)"))
                .hasMessage("Assigning NULL to non-null MERGE target table column");
```

Reply:
Good idea -- added.  I needed to change the `TRY(5/0)` into `CAST(TRY(5/0) AS VARCHAR)` to get the types to agree.